### ISSUE TITLE
[Rule] Go: use-filepath-join is not a very good rule

### ISSUE THREAD

**Issue by itaranto** [returntocorp/sgrep-rules#2426]
**Describe the bug**

The [use-filepath-join](https://github.com/returntocorp/semgrep-rules/blob/develop/go/lang/correctness/use-filepath-join.yaml) suggests to *always* use `filepath.Join` which is not entirely correct, `filepath` is meant to be used with filesystem paths and `path` is meant to be used with `/` separated paths (like URLs).

**To Reproduce**

```go
package main

import (
	"fmt"
	"net/url"
	"path"
)

func main() {
	url, err := url.Parse("http://foo:666/bar")
	if err != nil {
		panic(err)
	}

	fmt.Println(path.Join(url.Path, "baz"))
}
```

**Expected behavior**

I'm not sure what would be the expected behavior here, I believe it's very difficult to tell if `path.Join` is being used on a filesystem path or not. I'd rather not have this rule at all since it can be misleading.

**Priority**

How important is this to you?
- [ ] P0: blocking me from making progress
- [ ] P1: this will block me in the near future
- [x] P2: annoying but not blocking me

**Comment by LewisArdern** [comment]
Thanks for your report! Maybe we could do something like this? https://semgrep.dev/playground/s/lewisardern:use-filepath-join-copy?editorMode=advanced in general I agree this rule is not very nice -- but it definitely serves a purpose in file land. 

We could either do it:
1. Try to just remove anything with a HTTP request 
2. Only report when its within another `filepath.$SOMETHING(....,<... file.Join(...)...>,...)` or `file.$SOMETHING` etc.

**Comment by dgutson** [comment]
CC @qequ

**Comment by itaranto** [comment]
> Thanks for your report! Maybe we could do something like this? https://semgrep.dev/playground/s/lewisardern:use-filepath-join-copy?editorMode=advanced in general I agree this rule is not very nice -- but it definitely serves a purpose in file land.
> 
> We could either do it:
> 
>     1. Try to just remove anything with a HTTP request
> 
>     2. Only report when its within another `filepath.$SOMETHING(....,<... file.Join(...)...>,...)` or `file.$SOMETHING` etc.

Yes, something like this could work. My concern is that paths are just strings, I don't see a reliable way to identify paths, i.e. if they are FS paths or `/` separated paths.

**Comment by LewisArdern** [comment]
Sorry @itaranto  for the delayed response. I was on holiday. We could turn this into a taint rule to remove the noise, but in general it's an `correctness` rule and it should not be used for security purposes. There's a few considerations:
1. We change this to be a taint mode rule for net/http sources
2. We change this to be a normal rule where the function argument type string enters join 
3. We try to remove as much false positives as possible e.g. like i suggested above

cc @enncoded

**Comment by itaranto** [comment]
> Sorry @itaranto for the delayed response. I was on holiday.

No worries.

>   We change this to be a taint mode rule for net/http sources
>   We change this to be a normal rule where the function argument type string enters join
>   We try to remove as much false positives as possible e.g. like i suggested above

I think this makes a lot of sense!

**Comment by LewisArdern** [comment]
Assigning to @enncoded they will take it to the finishline :)

**Comment by enncoded** [comment]
We decided to go with the rule that Lewis posted earlier

**Comment by enncoded** [comment]
Sorry, this totally slipped my mind. Merging the change now.

**Comment by enncoded** [comment]
Merged! Gonna close this now. Feel free to reopen if y'all have anymore issues with this rule. Thanks for filing an issue!
